teps/0002-custom-tasks.md

```diff
@@ -0,0 +1,392 @@
+# TEP-0002: Enable Custom Tasks
+
+aka non-Pod Tasks, "duck-typed" Tasks 🦆
+
+Original Google Doc proposal, visible to members of tekton-dev@: https://docs.google.com/document/d/10nQSeIse7Ld4fLg4lhfgUmNKtewfaFNET3zlMdRnBuQ/edit
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Requirements](#requirements)
+- [Proposal](#proposal)
+  - [Validation](#validation)
+  - [Status Reporting](#status-reporting)
+  - [Pipeline Integration](#pipeline-integration)
+  - [Initial Update Timeout](#initial-update-timeout)
+  - [Parameter Passing](#parameter-passing)
+  - [Result Reporting](#result-reporting)
+  - [Cancellation](#cancellation)
+  - [Timeout](#timeout)
+  - [CLI and Dashboard Support](#cli-and-dashboard-support)
+  - [Results API](#results-api)
+  - [Changes to Triggers](#changes-to-triggers)
+  - [User Stories (optional)](#user-stories-optional)
+    - [Story 1](#story-1)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Test Plan](#test-plan)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Infrastructure Needed (optional)](#infrastructure-needed-optional)
+<!-- /toc -->
+
+## Summary
+
+Enable integrators to define new "custom" Task types as CRDs, which can be run
+by creating a new `Run` object (a CRD type which Tekton will define and
+own). Custom Task authors must also implement a reconciling controller which
+watches for creations of `Run` objects which reference their type, and
+which eventually updates its status to signal task completion.
+
+To achieve this, Tekton will define a new type, `Run`, first in
+`v1alpha1`, with the intention of iterating and, if there's support, bringing
+this functionality to `v1beta1` and eventually `v1`. Tekton will also
+implement support for this new type in the PipelineRun CRD controller, as well
+as document the contract for integrating with this new type, and providing
+examples and Go helper methods that integrators can consume if their
+controllers are implemented in Go.
+
+## Motivation
+
+Allow integrators to implement new functionality that doesn't directly map to
+Tekton's current idea of a `Task` and `TaskRun` (i.e., a collection of containers
+running in a `Pod` on a compute node).
+
+Some examples include (non-exhaustive):
+
+* wait for a period of time without having to schedule a container that just sleeps, wasting compute resources
+* wait for an external event to occur, e.g., an approval event signal
+* execute some operation outside of the cluster (e.g., a cloud build service, a macOS/mobile build farm) and wait for its execution to complete
+* execute another (sub-)Pipeline and wait for it to complete
+* enable matrix parameter expression -- e.g., succinctly express a Task that executes with args A,B,C (defined at config-time)
```

Comment:
Why the limitation that the matrix must be defined at config time?
It would be useful to have the matrix defined by the output of a previous task.

Reply:
🤷‍♂️  The point is, it's up to you, the Custom Task author, to define the behavior! 

I really only even mentioned it to contrast it with "looping tasks" that dynamically repeat behavior until some runtime condition is met. "Param-defined matrix tasks" like you describe is kind of a middle ground, where the Task is defined at config-time, but the actual parameters are only known at runtime. Could be a cool thing to prototype as a Custom Task amirite 😉